core/src/main/java/org/apache/accumulo/core/client/lexicoder/BigDecimalLexicoder.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.client.lexicoder;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+
+import org.apache.accumulo.core.clientImpl.lexicoder.FixedByteArrayOutputStream;
+
+/**
+ * A lexicoder to encode/decode a BigDecimal to/from bytes that maintain its native Java sort order.
+ *
```

Comment:
I think the biggest problem with the native Java sort order, is that it's non-deterministic (for items that differ only in scale). Lexicoder orderings should be deterministic. So, we'd need to use a comparator that produces the same output, no matter the order of the input, and have the lexicoder be implemented to preserve that ordering. Otherwise, the user experience could be very confusing. For example, the string lexical ordering of "2.0" and "2.00" would place "2.0" before "2.00". However, the comparator for BigDecimal is perfectly happy to order "2.00" before "2.0", because their `compareTo` value is `0`, even though their `equals` value is `false`.

Reply:
Shall we have the LexiCoder implement the Comparator interface or write a different Comparator for BigDecimals? Thanks

Reply:
I'm not sure either are strictly necessary. What matters is that the lexical ordering of the encoded form by this Lexicoder is deterministic, in numerical order first, then in lexical order by precision. So, `new BigDecimal(new BigInteger("203"), 2)` should sort lexically prior to `new BigDecimal(new BigInteger("2030"), 3)`, for example.

The main reason a Comparator is needed is to test that the ordering is preserved in both directions. So at the very least, a better comparator will be needed for testing. It may be sufficient to use `Comparator.naturalOrder().thenComparingInt(BigDecimal::precision)`, but I don't know if there are any edge cases where this breaks.

As for the javadoc, rather than stating that we're maintaining its native Java sort order, we should specify that we're maintaining its numerical order and precision, so that numerically equivalent values, the ones with the least precision are ordered first. When this lexicoder is reversed, the opposite should be true.